synapse/handlers/sync.py

```diff
@@ -485,22 +495,60 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
         # TODO(mjark) Check for new redactions in the state events.
 
         with Measure(self.clock, "compute_state_delta"):
+
+            types = None
+            member_state_ids = {}
+            lazy_load_members = sync_config.filter_collection.lazy_load_members()
+
+            if lazy_load_members:
+                # We only request state for the members needed to display the
+                # timeline:
+
+                types = [
+                    (EventTypes.Member, state_key)
+                    for state_key in set(
+                        event.sender  # FIXME: we also care about invite targets etc.
+                        for event in batch.events
+                    )
+                ]
+
+                # We can't remove redundant member types at this stage as it has
+                # to be done based on event_id, and we don't have the member
+                # event ids until we've pulled them out of the DB.
+
+                if not types:
+                    # an optimisation to stop needlessly trying to calculate
+                    # member_state_ids
+                    #
+                    # XXX: i can't remember what this trying to do. why would
+                    # types ever be []? --matthew
+                    lazy_load_members = False
+
+                types.append((None, None))  # don't just filter to room members
+
             if full_state:
                 if batch:
                     current_state_ids = yield self.store.get_state_ids_for_event(
-                        batch.events[-1].event_id
+                        batch.events[-1].event_id, types=types
                     )
 
                     state_ids = yield self.store.get_state_ids_for_event(
-                        batch.events[0].event_id
+                        batch.events[0].event_id, types=types
                     )
+
                 else:
                     current_state_ids = yield self.get_state_at(
-                        room_id, stream_position=now_token
+                        room_id, stream_position=now_token, types=types
                     )
 
                     state_ids = current_state_ids
 
+                if lazy_load_members:
+                    member_state_ids = {
```

Comment:
why are we doing this here and at line 582, rather than just doing it in `_calculate_state` based on a `lazy_load_members` flag?

Reply:
oh, is it because of the TODO at line 578 - this is supposed to do something more intelligent?

Reply:
yes. on the https://github.com/matrix-org/synapse/pull/3331/files branch the second branch evolves to filter out the redundant members to stop repeatedly spamming lazy-loaded members to the client in each and every /sync.